**Multivariate Regression**

We are going to perform multivariate regression:

Independent variables: VADER Raw (x1), Adjectives & Vader (x2), KNN (x3)

Dependent variables: Review star rating

Star rating = B0 + B1x1 + B2x2 + B3x3 + error term

In [4]:
import pandas as pd
import statsmodels.api as sm

df = pd.read_csv('Module3_Stars_Sentiment.csv', encoding='cp1252') # encoding for CSVs exported from Excel on Windows
X = df[['VADER_stars', 'JJ_VADER_stars', 'KNN_stars']]
y = df['Review Stars']

X = sm.add_constant(X)  # Adds a constant term to the predictor variables
model = sm.OLS(y, X).fit()  # Fit the OLS regression model
print(model.summary())  # Print the summary of the regression results

                            OLS Regression Results                            
Dep. Variable:           Review Stars   R-squared:                       0.497
Model:                            OLS   Adj. R-squared:                  0.485
Method:                 Least Squares   F-statistic:                     43.15
Date:                Tue, 06 Oct 2026   Prob (F-statistic):           1.85e-19
Time:                        21:53:09   Log-Likelihood:                -204.86
No. Observations:                 135   AIC:                             417.7
Df Residuals:                     131   BIC:                             429.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const             -0.8868      0.385     -2.

**Interpreting Coefficients**

JJ_VADER_stars and KNN_stars are statistically significant positive predictors of Review Stars since they have positive coefficients and p-values less than 0.05, while VADER_stars is not statistically significant (if considering p-value of 0.05). Holding the other predictors constant, a one-unit increase in JJ_VADER_stars is associated with a 0.57-star increase in Review Stars, while a one-unit increase in KNN_stars is associated with a 0.48-star increase. As we mentioned, VADER_stars relationship with Review Stars is not statistically significant (p = 0.123).

The sentiment method that best predicts star rating depends on what we mean by best.

If we're looking at effect size:
JJ_VADER is best, because it has the largest coefficient (0.569).

If we're looking at statistical reliability:
KNN is best, because it has the largest t-statistic (4.549) and smallest p-value (<0.001)

**Assessing Model Fit**

As we can see from the OLS Regression Results, our model has an R² of 0.497 and an adjusted R² of 0.485. This means that our 3 predictors explain approximately 48.5% of the variation in Review Stars after accounting for the number of predictors. The small difference between R² and adjusted R² indicates that the model maintains strong explanatory power after accounting for the number of predictors, suggesting limited overfitting.